File: Java/LeetCode/Find Players With Zero or One Losses.java
// https://leetcode.com/problems/find-players-with-zero-or-one-losses

class Solution {

    public List<List<Integer>> findWinners(int[][] matches) {
        Set<Integer> winners = new TreeSet<>();
        Map<Integer, Integer> playerLosses = new TreeMap<>();

        for (int[] match : matches) {
            int winner = match[0];
            winners.add(winner);

            int loser = match[1];
            int losses = playerLosses.getOrDefault(loser, 0);
            playerLosses.put(loser, losses+1);
        }

        List<Integer> playersWithZeroLosses = new ArrayList<>();
        for (Integer player : winners) {
            if (!playerLosses.containsKey(player)) {
                playersWithZeroLosses.add(player);
            }
        }

        List<Integer> playersWithOneLoss = new ArrayList<>();
        for (Integer player : playerLosses.keySet()) {
            if (playerLosses.get(player) == 1) {
                playersWithOneLoss.add(player);
            }
        }

        List<List<Integer>> answer = new ArrayList<>();
        answer.add(playersWithZeroLosses);
        answer.add(playersWithOneLoss);

        return answer;
    }

}

File: Kotlin/LeetCode/Find Players With Zero or One Losses.kt
// https://leetcode.com/problems/find-players-with-zero-or-one-losses

class Solution {
    
    fun findWinners(matches: Array<IntArray>): List<List<Int>> {
        val winners = sortedSetOf<Int>();
        val playerLosses = sortedMapOf<Int, Int>();

        for (match in matches) {
            val winner = match[0];
            winners.add(winner);

            val loser = match[1];
            val losses = playerLosses.getOrDefault(loser, 0);
            playerLosses.put(loser, losses+1);
        }

        val playersWithZeroLosses = mutableListOf<Int>();
        for (player in winners) {
            if (!playerLosses.containsKey(player)) {
                playersWithZeroLosses.add(player);
            }
        }

        val playersWithOneLoss = mutableListOf<Int>();
        for (player in playerLosses.keys) {
            if (playerLosses.get(player) == 1) {
                playersWithOneLoss.add(player);
            }
        }

        val answer = mutableListOf(playersWithZeroLosses, playersWithOneLoss)

        return answer;
    }

}
